src/backend/cdb/cdbmutate.c

```diff
@@ -1014,6 +1014,18 @@ shareinput_mutator_xslice_2(Node *node, PlannerInfo *root, bool fPop)
 					break;
 			}
 		}
+
+		/*
+		 * Remove writer gang for the shareinput scan readers in case
+		 * of cross-slice modifying CTE
+		 */
+		if (sisc->cross_slice &&
+			root->parse->hasModifyingCTE &&
+			currentSlice->gangType == GANGTYPE_PRIMARY_WRITER &&
+			!plan->lefttree)
+		{
+			currentSlice->gangType = GANGTYPE_PRIMARY_READER;
+		}
```

Comment:
Well, there is a case: when in **one slice we have producer and consumer**, and the **third** consumer is in another slice. In this case the first consumer will have `sisc->cross_slice` `true`, and it will overwritte the correct writer gangtype.
Before patch:
```
create table t_hashed(i int, j int);
create table t2 (i int, j int);

explain (slicetable,costs off) with cte as (insert into t_hashed select i, i*2 from generate_series(1,5) i  returning *) 
select * from cte union all select * from cte union all (select t2.j, cte.i from cte join t2 using (i));
                                  QUERY PLAN                                  
------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)
   ->  Append
         ->  Subquery Scan on cte
               ->  Shared Scan (share slice:id 1:0)
                     ->  Insert on t_hashed
                           ->  Redistribute Motion 1:3  (slice2; segments: 1)
                                 Hash Key: i.i
                                 ->  Function Scan on generate_series i
         ->  Subquery Scan on cte_1
               ->  Shared Scan (share slice:id 1:0)
         ->  Hash Join
               Hash Cond: (t2.i = cte_2.i)
               ->  Seq Scan on t2
               ->  Hash
                     ->  Redistribute Motion 3:3  (slice3; segments: 3)
                           Hash Key: cte_2.i
                           ->  Subquery Scan on cte_2
                                 ->  Shared Scan (share slice:id 3:0)
 Optimizer: Postgres-based planner
 Slice 0: Dispatcher; root 0; parent -1; gang size 0
 Slice 1: Primary Writer; root 0; parent 0; gang size 3
 Slice 2: Singleton Reader; root 0; parent 1; gang size 1; segment 1
 Slice 3: Reader; root 0; parent 1; gang size 3
(23 rows)
```
With patch:
```
explain (slicetable,costs off) with cte as (insert into t_hashed select i, i*2 from generate_series(1,5) i  returning *) 
select * from cte union all select * from cte union all (select t2.j, cte.i from cte join t2 using (i));
                                  QUERY PLAN                                  
------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)
   ->  Append
         ->  Subquery Scan on cte
               ->  Shared Scan (share slice:id 1:0)
                     ->  Insert on t_hashed
                           ->  Redistribute Motion 1:3  (slice2; segments: 1)
                                 Hash Key: i.i
                                 ->  Function Scan on generate_series i
         ->  Subquery Scan on cte_1
               ->  Shared Scan (share slice:id 1:0)
         ->  Hash Join
               Hash Cond: (t2.i = cte_2.i)
               ->  Seq Scan on t2
               ->  Hash
                     ->  Redistribute Motion 3:3  (slice3; segments: 3)
                           Hash Key: cte_2.i
                           ->  Subquery Scan on cte_2
                                 ->  Shared Scan (share slice:id 3:0)
 Optimizer: Postgres-based planner
 Slice 0: Dispatcher; root 0; parent -1; gang size 0
 Slice 1: Reader; root 0; parent 0; gang size 3
 Slice 2: Singleton Reader; root 0; parent 1; gang size 1; segment 0
 Slice 3: Reader; root 0; parent 1; gang size 3
(23 rows)
```
I think the most valid solution is no ensure that current SISC consumer is definetly in a different slice from  the producer.